keps/sig-node/1664-node-ips/README.md

```diff
@@ -0,0 +1,491 @@
+possible that other components might attempt to update the field
+incorrectly in the future.
+
+#### `NodeAddressType` semantics
+
+The difference between `InternalIP` and `ExternalIP` node addresses is
+not well-defined except in public clouds, and the two are used
+inconsistently outside of that environment (though for the most part,
+`ExternalIP` is used to mean "involves address rewriting from a public
+IP to a private IP", and `InternalIP` is used for anything else).
+
+### Goals
+
+- Add `Node.Status.Hostname` and `Node.Status.IPs` and deprecate
+  `Node.Status.NodeAddresses`
+
+- Clarify the expectations of cloud providers and
+  cloud-controller-manager with respect to single-stack IPv6 and
+  dual-stack clusters.
+
+  - Do better error handling when, eg, trying to bring up a single-stack
+    IPv6 cluster on a cloud that only supports IPv4.
+
+- Extend the bare-metal node address code to allow for dual-stack
+
+- Clarify the expected behavior of `--node-ip` and extend it to handle
+  dual-stack clusters.
+
+- Clarify the semantics of "internal" vs "external" node IPs and ensure
+  that all cloud providers are using the distinction correctly.
+
+### Non-Goals
+
+- Requiring cloud providers to implement IPv6 / dual-stack.
+
+### Maybe Goals
+
+- Some people want the ability to set `ExternalIP` addresses on
+  bare-metal nodes ([kubernetes #42125]). This is potentially distinct
+  from being able to set a dual-stack pair of `InternalIP` addresses.
+
+[kubernetes #42125](https://github.com/kubernetes/kubernetes/issues/42125)
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system.  The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+#### Story 2
+
+### Notes/Constraints/Caveats (optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above.
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal and how do we mitigate.  Think broadly.
+For example, consider both security and how this will impact the larger
+kubernetes ecosystem.
+
+How will security be reviewed and by whom?
+
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+The `NodeAddress` of type `NodeHostName` will be moved to a new
+`Node.Status.Hostname` field.
```

Comment:
What is the meaning of this field?  Is it the hostname as the node knows itself (e.g. $(hostname) or $(hostname -f)) or is it any DNS that can be resolved to any one of the node's IPs?  Or is it the name you get when you reverse a particular node IP? Where must it be DNS resolvable from?

Reply:
What is this used for?

Reply:
I don't know the answer to any of those questions... currently `Addresses` contains a single element of type `NodeHostName` and one or more IP/DNS elements. This proposal was just splitting the hostname out into a separate field from the IP/DNS stuff.

I had an "UNRESOLVED" section noting that I wasn't sure whether the hostname was currently required or optional. I guess I should check if it's even *used*...

Reply:
added a bunch of notes about what NodeHostName is used for...